### DISCUSSION TITLE
How to access the dbxrefs in the subfeatures?

### DISCUSSION THREAD

**Discussion by Chi-HsienChang** [GMOD/jbrowse-components#3603] [Q&A] [answered]
Hi JBrowse2 developers,

Our group has started working on a plugin to handle linkouts for dbxrefs in features and subfeatures.  
The specific use case is to use this with NCBI generated gene sets so that users can link out to the different NCBI resource pages.  We are done well in getting the linkouts working for dbxrefs at the feature level, but we are having problems figuring out how to access the dbxrefs in the subfeatures.  Below is a link to the GitHub repo for the plugin, and any advice you might have on accessing subfeatures would be greatly appreciated.

https://github.com/Chi-HsienChang/jbrowse-plugin-template

Thank you!

**Comment by cmdcolin** [comment, marked as answer]
@Chi-HsienChang How is your plugin being used in e.g. your config file? I see it defines custom jexl functions but how are they used

**Reply by Chi-HsienChang** [reply]
Hi @cmdcolin,

I use the plugin by adding the below setting to jbrowse-plugin-template/.jbrowse/config.json manually:
      "formatDetails": {
        "feature": "jexl:{dbxref:linkout_by_dbxref(feature), description:linkout_to_ncbi(feature)}"
      }

In addition, I study about how to automatically add the setting to the user's config.json by the plugin now.

Thank you!

**Reply by cmdcolin** [reply]
Thanks for this info

- What is an example GFF e.g. where is the Genbank stored? I would recommend not JSON.stringifying things probably because it removes the structure of the feature object. It may work ok, but if you have knowledge of where the Genbank field is stored on the feature object, you can access that directly instead of stringifying the whole feature object and searching inside the string

- Do you want to display "on the parent feature" the details of the subfeatures? If so, then your approach may be fine, you can access feature.subfeatures (you do this in your plugin, but you also didn't explain what error you ran into doing this). If not, then you can also consider adding links to the subfeatures directly, but there was a bug related to this that is being fixed here https://github.com/GMOD/jbrowse-components/pull/3608

**Reply by cmdcolin** [reply]
I also made another issue for your request to make plugins be able to set a custom default value here https://github.com/GMOD/jbrowse-components/issues/3609

**Comment by Chi-HsienChang** [comment]
Hi @cmdcolin ,

Thank you for your guidance and help!

1. I add the example GFF file to the path ./src/file.gff

    I want to access that directly instead of stringifying the whole feature object and searching inside the string. However, I fail it.

2. I want to use subfeatures.name or subfeatures.id (https://d.pr/i/1qRvcI) to access the variable, it does not work for me.
      For example:

    "formatDetails": {
        "feature": "jexl:{NCBI_Gene:linkout_gene(feature)}",
        "subfeatures": "jexl:{name:'<a href=https://google.com/?q='+subfeatures.name+'>'+subfeatures.name+'<slash&a>'}"
      }
			
	"feature" works for me,  but "subfeatures" does not work for me.
	I want to add links to the subfeatures directly.
	
3.  I am not sure whether the plugin is able to automatically add the setting:
	    "formatDetails": { "feature": "jexl:{NCBI_Gene:linkout_gene(feature)}" }  
        to the user's config.json on JBrowse2 desktop or web versions.
			
4. I want to ask whether I need to clone https://github.com/GMOD/jbrowse-components.git to develop our linkout plugin. Now, I just 
     add jexl functions to the plugin's src/index.ts.
     If I want to publish the plugin in the future whether it necessary to clone the jbrowse-components to develop it?

Thank you!

**Reply by cmdcolin** [reply]
1. see end of this post for example that I tested using this GFF
2. subfeatures will start working after i release the next version, there was an unfortunate bug! in this example though, the variable name is still named "feature" inside the jexl callback e.g. it may look something like `"subfeatures": "jexl:{name:'<a href=[https://google.com/?q='+subfeatures.name+'>'+feature.name+](https://google.com/?q=%27+feature.name+%27%3E%27+feature.name+)'<slash&a>'}"`
3. i don't know how to do this off the top of my head, but made a issue to try to track this here https://github.com/GMOD/jbrowse-components/issues/3609
4. the process for developing plugins can be varied. I think running a github clone of https://github.com/GMOD/jbrowse-components is wise. you will get better error messages running the "dev version" of jbrowse web (e.g. running yarn start in products/jbrowse-web with your plugin loaded). Using the jbrowse-plugin-template is probably good also. This is one guide for developing a plugin https://jbrowse.org/jb2/docs/tutorials/simple_plugin/ see also "no build plugins" which are more lightweight and are nice for just adding extra jexl functions https://jbrowse.org/jb2/docs/tutorials/no_build_plugin_tutorial/


## Working example finding the right Genbank Dbxref:



config.json snippet

```
{
      "type": "FeatureTrack",
      "trackId": "genes",
      "name": "genes",
      "formatDetails": {
        "subfeatures": "jexl:{name:linkout_gene(feature)}"
      },
      "adapter": {
        "type": "Gff3TabixAdapter",
        "gffGzLocation": { "uri": "out.sort.gff.gz" },
        "index": {
          "location": { "uri": "out.sort.gff.gz.tbi" },
          "indexType": "TBI"
        }
      },
      "assemblyNames": ["bombus"]
    }
```

plugin snippet with the linkout_gene function

```
 pluginManager.jexl.addFunction('linkout_gene', feature => {
        let dbxref = feature.dbxref
          ?.find(f => f.startsWith('Genbank:'))
          ?.replace('Genbank:', '')

        if (!dbxref) {
          return feature.name // return name like normal
        }
        // name is replaced with a dbxref link, keeps the feature.name inside the link text or uses dbxref if no feature.name
        return `<a href=https://www.ncbi.nlm.nih.gov/gene/?term=${dbxref}>${feature.name||dbxref}</a>`
      })
```

**Reply by cmdcolin** [reply]
v2.4.2 released now :) above config snippet with formatDetails->subfeatures will work in that version

**Reply by Chi-HsienChang** [reply]
Hi @cmdcolin,
Your responses are very helpful to me. I successfully accesse the subfeatures now. Thank you very much!

**Reply by Chi-HsienChang** [reply]
Hi @cmdcolin,

I have an another question:

I want to use the similar way to instead of JSON.stringify(feature.dbxref), but it does not work (subfeatures works for me).
I try several ways, I don't know how to fix it.

```
`"formatDetails": {
        "feature": "jexl:{dbxref:linkout_GeneID(feature)}",
        "subfeatures": "jexl:{dbxref:linkout_Genbank(feature)}"
      },
```

```
pluginManager.jexl.addFunction('linkout_GeneID', (feature: any) => {
      let dbxref = feature.dbxref
        ?.find((f: any) => f.startsWith('GeneID:'))
        ?.replace('GeneID:', '')
      if (!dbxref) {
        return feature.name 
      }
      return `<a href=https://www.ncbi.nlm.nih.gov/gene/?term=${dbxref}>GeneID:${feature.name||dbxref}</a>`
    })
```


Thank you!

**Reply by cmdcolin** [reply]
potential fix @Chi-HsienChang  edit: see https://github.com/GMOD/jbrowse-components/discussions/3603#discussioncomment-5470742

```
pluginManager.jexl.addFunction('linkout_GeneID', feature => {
        if (feature.dbxref) {
          const ref = Array.isArray(feature.dbxref)
            ? feature.dbxref
            : [feature.dbxref]
          let dbxref = ref
            ?.find(f => f.startsWith('GeneID:'))
            ?.replace('GeneID:', '')
          if (!dbxref) {
            return feature.name
          }
          return `<a href=https://www.ncbi.nlm.nih.gov/gene/?term=${dbxref}>GeneID:${
            feature.name || dbxref
          }</a>`
        }
        return feature.name
      })
```

**Reply by cmdcolin** [reply]
you can see this code checks if feature.dbxref is an array, and if so, it uses it normally, otherwise it turns it into an array (with one element, itself). this is sort of awkward but is needed due to how the gff is parsed...if it is a single entry, it is a plain string, if it is multiple entries, it is an array of strings

in our other data parsing modules like @gmod/vcf, we make "everything an array by default" but with gff it is sometimes a string and sometimes an array

**Reply by Chi-HsienChang** [reply]
Hi @cmdcolin ,

Thank you for your guidance, I understand, thank you!

**Comment by cmdcolin** [comment]
@Chi-HsienChang Here is an alternative solution that handles multiple dbxref types in a single callback, and importantly handles MULTIPLE dbxrefs in a single feature. before, the usage of "find" would find the single dbxref type we were looking for and ignore others, but then the others wouldn't get printed in the feature details.

This new callback will properly handle multiple

Example


```
pluginManager.jexl.addFunction('linkout', feature => {
        // no dbxref found, so return empty string
        if (!feature.dbxref) {
          return ''
        }
        const dbxrefs = Array.isArray(feature.dbxref)
          ? feature.dbxref
          : [feature.dbxref]
        return dbxrefs.map(dbxref => {
          // customized link for Genbank dbxref
          if (dbxref.startsWith('Genbank:')) {
            const ref = dbxref.replace('Genbank:', '')
            return `<a href=https://www.ncbi.nlm.nih.gov/gene/?term=${ref}>${dbxref}</a>`
          }
          // customized link for GeneID dbxref
          else if (dbxref.startsWith('GeneID:')) {
            const ref = dbxref.replace('GeneID:', '')
            return `<a href=https://www.ncbi.nlm.nih.gov/gene/?term=${ref}>${dbxref}</a>`
          }
          // no link, just plaintext returned
          return dbxref
        })
      })
```


Config


```
"formatDetails": {
        "feature": "jexl:{dbxref:linkout(feature)}",
        "subfeatures": "jexl:{dbxref:linkout(feature)}"
      }
```

Screenshot

![image](https://user-images.githubusercontent.com/6511937/228673942-71259783-dfac-49de-af52-16717b00ce6e.png)


You can see on the subfeature it correctly has multiple dbxref links for that entry

**Reply by Chi-HsienChang** [reply]
Hi @cmdcolin, 

It works for me, thank you so so much!
